perf_bench_test.go

```diff
@@ -24,127 +24,567 @@ package main
-			tStart := time.Now()
-			go serveVM(t, tStart, vmIDString, imageName, &vmGroup, isSyncOffload)
+		log.Infof("Current RPS: %d", rps)
 
-			vmID = (vmID + 1) % *vmNum
-		case <-done:
-			ticker.Stop()
+		tStart := time.Now()
+		latencyCh := make(chan LatencyStat)
+		go latencyMeasurement(t, vmNum, images, &isProfile, latencyCh)
+
+		err := profiler.Run()
+		require.NoError(t, err, "Run profiler returned error")
+		go profileControl(&isProfile, profiler)
+
+		for remainingRequests > 0 {
+			if tickerT := <-ticker.C; !tickerT.IsZero() {
+				vmGroup.Add(1)
+				remainingRequests--
+
+				imageName := images[vmID%len(images)]
+
+				go loadVM(t, &vmGroup, vmID, requestID, imageName, isSyncOffload, &isProfile, &invokExecTime, &realRequests)
+				requestID++
+
+				vmID = (vmID + 1) % vmNum
+			}
 		}
+		ticker.Stop()
+		vmGroup.Wait()
+		log.Debugf("All VM returned in %d Milliseconds", time.Since(tStart).Milliseconds())
+		latencies := <-latencyCh
+		log.Debugf("Mean Latency: %f, Tail Latency: %f", latencies.meanLatency, latencies.tailLatency)
+
+		// Collect results
+		serveMetric.MetricMap[avgExecTime] = float64(invokExecTime) / float64(realRequests)
+		serveMetric.MetricMap[rpsPerCore] = float64(realRequests) / (profiler.GetCoolDownTime() - profiler.GetWarmUpTime())
+		if cores > vmNum {
+			serveMetric.MetricMap[rpsPerCore] /= float64(vmNum)
+		} else {
+			serveMetric.MetricMap[rpsPerCore] /= float64(cores)
+		}
+		result, err := profiler.GetResult()
+		profiledCores := profiler.GetCores()
+		log.Debugf("%d cores are recorded: %v", len(profiledCores), profiledCores)
+		require.NoError(t, err, "Stopping profiler returned error: %v", err)
+		for eventName, value := range result {
+			log.Debugf("%s: %f", eventName, value)
+			serveMetric.MetricMap[eventName] = value
+		}
+		log.Debugf("%s: %f", avgExecTime, serveMetric.MetricMap[avgExecTime])
+		log.Debugf("%s: %f", rpsPerCore, serveMetric.MetricMap[rpsPerCore])
+		profiler.PrintBottlenecks()
+
+		// if tail latency violates the contraints that it should be less than 5x service time,
+		// it returns the metric before tail latency violation.
+		if latencies.tailLatency > threshold {
+			require.NotNil(t, pmuMetric, "The tail latency of first round is larger than the constraint")
+			return pmuMetric.MetricMap
+		}
+
+		pmuMetric = serveMetric
 	}
 
-	done <- true
-	vmGroup.Wait()
+	return pmuMetric.MetricMap
 }
 
-///////////////////////////////////////////////////////////////////////////////
-////////////////////////// Auxialiary functions below /////////////////////////
-///////////////////////////////////////////////////////////////////////////////
-func serveVM(t *testing.T, start time.Time, vmIDString, imageName string, vmGroup *sync.WaitGroup, isSyncOffload bool) {
+// Goroutine function: serve VM, record real RPS and exection time
+func loadVM(t *testing.T, vmGroup *sync.WaitGroup, vmID, requestID int, imageName string, isSyncOffload bool, isProfile *bool, execTime, realRequests *int64) {
```

Comment:
rename: loadVMs

please split the long line into few (one argument per line)

Reply:
Done